manager/orchestrator/restart/restart.go

```diff
@@ -190,6 +192,83 @@ func (r *Supervisor) Restart(ctx context.Context, tx store.Tx, cluster *api.Clus
 	return nil
 }
 
+// TaskRestartDelay calculates and returns the next delay based t's RestartPolicy
+func (r *Supervisor) TaskRestartDelay(ctx context.Context, t *api.Task) (*time.Duration, error) {
+	var restartDelay time.Duration
+	backoff := &api.BackoffPolicy{
+		Base:   defaults.Service.Task.Restart.Backoff.Base,
+		Factor: defaults.Service.Task.Restart.Backoff.Factor,
+		Max:    defaults.Service.Task.Restart.Backoff.Max,
+	}
+	if t.Spec.Restart != nil {
+		if t.Spec.Restart.Backoff != nil {
+			if a, b := gogotypes.DurationFromProto(t.Spec.Restart.Backoff.Base); b == nil && a >= 0 {
+				backoff.Base = t.Spec.Restart.Backoff.Base
+			}
+			if a, b := gogotypes.DurationFromProto(t.Spec.Restart.Backoff.Factor); b == nil && a >= 0 {
+				backoff.Factor = t.Spec.Restart.Backoff.Factor
+			}
+			if a, b := gogotypes.DurationFromProto(t.Spec.Restart.Backoff.Max); b == nil && a >= 0 {
+				backoff.Max = t.Spec.Restart.Backoff.Max
+			}
+		} else if t.Spec.Restart.Delay != nil {
+			// Use fixed restart delay
+			var err error
+			restartDelay, err = gogotypes.DurationFromProto(t.Spec.Restart.Delay)
+			if err != nil {
+				log.G(ctx).WithError(err).Error("invalid restart delay; using default")
+				restartDelay, err = gogotypes.DurationFromProto(defaults.Service.Task.Restart.Delay)
+				if err != nil {
+					return nil, errors.New("error parsing RestartPolicy")
+				}
+			}
+			return &restartDelay, nil
+		}
+	}
+
+	// Use jittered binary exponential backoff delay
+	var base, factor, max time.Duration
+	var err error
+	if base, err = gogotypes.DurationFromProto(backoff.Base); err != nil {
+		return nil, errors.New("error parsing backoff Base")
+	}
+	if factor, err = gogotypes.DurationFromProto(backoff.Factor); err != nil {
+		return nil, errors.New("error parsing backoff Factor")
+	}
+	if max, err = gogotypes.DurationFromProto(backoff.Max); err != nil {
+		return nil, errors.New("error parsing backoff Max")
+	}
+
+	var failures uint64
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	serviceID := t.ServiceID
+	tuple := orchestrator.SlotTuple{
+		Slot:      t.Slot,
+		ServiceID: t.ServiceID,
+		NodeID:    t.NodeID,
+	}
+	if r.historyByService[serviceID] != nil &&
+		r.historyByService[serviceID][tuple] != nil {
+		restartInfo := r.historyByService[serviceID][tuple]
+		failures = restartInfo.failuresSinceSuccess
+	} else {
+		// History is not set on the first run of a task
+		failures = 0
+	}
+
+	// Use floating point values to detect overflow.  math.Pow returns +Inf on overflow
+	backoffDuration := float64(base) + float64(factor)*math.Pow(2, float64(failures))
```

Comment:
Thanks for changing this to `math.Pow(2, float(failures))`. I know I was wrong about the bit shift being multiplication (it wasn't) but that's part of the reason it should have been changed.